include/flox/resolver/descriptor.hh

```diff
@@ -0,0 +1,182 @@
+
+/**
+ * @fn void from_json( const nlohmann::json        & jfrom
+ *                   ,       ManifestDescriptorRaw & desc
+ *                   )
+ * @brief Convert a JSON object to an @a flox::InputPreferences.
+ *
+ * @fn void to_json( nlohmann::json & jto, const ManifestDescriptorRaw & desc )
+ * @brief Convert an @a flox::resolver::ManifestDescriptorRaw to a JSON object.
+ */
+/* Generate to_json/from_json functions. */
+NLOHMANN_DEFINE_TYPE_NON_INTRUSIVE_WITH_DEFAULT( ManifestDescriptorRaw
+                                               , name
+                                               , version
+                                               , stability
+                                               , path
+                                               , absPath
+                                               , systems
+                                               , optional
+                                               , packageGroup
+                                               , packageRepository
+                                               , input
+                                               )
+
+
+/* -------------------------------------------------------------------------- */
+
+/**
+ * @brief Extend and remap fields from @a flox::resolver::PkgDescriptorRaw to
+ *        those found in a `flox` _manifest_.
+ */
+struct ManifestDescriptor {
+
+  public:
+
+    /** Match `name`, `pname`, or `pkgAttrName` */
+    std::optional<std::string> name;
+
+    /** Whether resoution is allowed to fail without producing errors. */
+    bool optional = false;
+
+    /** Named _group_ that the package is a member of. */
+    std::optional<std::string> group;
+
+    /** Match `version`. */
+    std::optional<std::string> version;
+
+    /** Match a semantic version range. */
+    std::optional<std::string> semver;
+
+    /** Match a subtree. */
+    std::optional<Subtree> subtree;
+
+    /** Only resolve for a given set of systems. */
+    std::optional<std::vector<std::string>> systems;
+
+    /** Match a catalog stability. */
+    std::optional<std::string> stability;
+
+    /** Match a relative attribute path. */
+    std::optional<flox::AttrPath> path;
+
+    /** Force resolution is a given input, _flake reference_, or file. */
+    std::optional<std::variant<nix::FlakeRef, std::string>> input;
+
+
+    ManifestDescriptor() = default;
+
+    explicit ManifestDescriptor( const ManifestDescriptorRaw & raw );
+
+
+    /** @brief Reset to default state. */
+    void clear();
+
+    /**
+     * @brief Fill a @a flox::pkgdb::PkgQueryArgs struct with preferences to
+     *        lookup packages.
+     *
+     * NOTE: This DOES NOT clear @a pqa before filling it.
+     * This is intended to be used after filling @a pqa with global preferences.
+     * @param pqa A set of query args to _fill_ with preferences.
+     * @return A reference to the modified query args.
```

Comment:
Thought: I find it hard to understand exactly how `fillPkgQueryArgs` fits in with the rest of the codebase, because it's not clear what the state of the `PkgQueryArgs` is before it gets modified here, and it's not clear what modifications might happen afterwards. And since there are so many different definitions of `fillPkgQueryArgs` it's a bit hard to understand how they all fit together. I wonder if at some point it might be more clear to split things up into something like query args + registry args + preferences args, with each one of those having a simpler lifecycle, and then there could be one place they were all pulled together into `PkgQueryArgs`